### PR TITLE
Fix: Error when scss file is empty #327

### PR DESCRIPTION
**Pull request by gongpeione**

### CODE REVIEW

**Comment by brandon93s** [comment]
Would like to understand why this is needed ([issue](https://github.com/parcel-bundler/parcel/issues/327#issuecomment-352614344)). This fixes things currently, but there is no guarantee that an empty string fixes it in future versions of `node-sass`. Ideally, we'd skip `node-sass` if this is truly needed.

**Comment by yeskunall** [comment]
@brandon93s while there is no use-case, it shouldn't break `parcel`'s bundling process. Similar issue has been brought up at webpack-contrib/sass-loader#403, and I believe this check should be in [src/Asset.js#parse](https://github.com/parcel-bundler/parcel/blob/master/src/Asset.js#L93) instead.

All assets extend from the base class, so we can avoid duplication by adding the check (or something along the lines of that) added by @gongpeione there.

**Comment by devongovett** [comment]
What is the error you're getting with an empty file?

**Comment by yeskunall** [comment]
Looks like the specific error message is

> No input specified: provide a file name a source string to process

... taken from the issue.

**Comment by gongpeione** [comment]
Yes like @yeskunall said, the error message is `No input specified: provide a file name a source string to process`, cause node-sass treat empty file like null
```javascript
if (options.data) {
    binding.render(options);
} else if (options.file) {
    binding.renderFile(options);
} else {
    cb({status: 3, message: 'No input specified: provide a file name or a source string to process' });
}
```

File: src/Asset.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ class Asset {
   async getDependencies() {
     await this.loadIfNeeded();
 
-    if (this.mightHaveDependencies()) {
+    if (this.mightHaveDependencies() && this.contents) {
```
Comment (yeskunall) [inline src/Asset.js]: `mightHaveDependencies` already checks for `this.contents`, so this additional check is not required.

File: src/Asset.js [hunk 2]
```diff
@@ -82,7 +82,11 @@ class Asset {
       .generateBundleName();
   }
 
-  mightHaveDependencies() {
+  async mightHaveDependencies() {
+    await this.loadIfNeeded();
+    if (!this.contents) {
```
Comment (yeskunall) [inline src/Asset.js]: If `this.contents` is passed along a "falsy" value, such as an empty string, then `mightHaveDependencies` will return `false`, which might cause unintended behaviour.

File: src/Asset.js (hunk 1 above)
Comment (gongpeione) [reply src/Asset.js]: Some Asset class like CSSAsset override `getDependencies` method or some others may override `getDependencies`, this is wanna double check if contents needs to be parsed, trying not to modifiy other Asset class.

File: src/Asset.js (hunk 2 above)
Comment (gongpeione) [reply src/Asset.js]: I am a little confused, this is indeed wants to skip `parse()` if `this.contents` is an empty string.

File: src/Asset.js (hunk 1 above)
Comment (devongovett) [reply src/Asset.js]: swap the order of these conditions so `mightHaveDependencies` doesn't run if `this.contents` is empty.

File: src/Asset.js [hunk 3]
```diff
@@ -83,6 +83,9 @@ class Asset {
   }
 
   mightHaveDependencies() {
+    if (!this.contents) {
+      return false;
+    }
```
Comment (devongovett) [inline src/Asset.js]: remove this check as it's already done above

**Pull request merged**